
package LifeSim;


message VectorDesc {

  repeated float x = 1;

}

message TransformDesc {
  required VectorDesc position = 1;
  required VectorDesc rotation = 2;
}




message SceneObjectDesc {

  extensions 1000 to max;

  enum Type {
		PLANE    = 0;
    BOX      = 1;
		SPHERE   = 2;
		CREATURE = 100;
		LIGHT    = 1000;
  }

  required Type type = 1;
  required string name = 2;

	// TODO : Refactor before release, rigid bodies and their states should be decoupled.
	//        This currently is used to specify the default positions of the
	//        rigid bodies.
	required TransformDesc transform = 3;

}

message ScenePlaneDesc {

  extend SceneObjectDesc {
		required ScenePlaneDesc scene_object = 1000;
	}

	required VectorDesc plane = 1;
  required float size = 2;
  required VectorDesc origin = 3;

}


message SceneBoxDesc {

  extend SceneObjectDesc {
    required SceneBoxDesc scene_object = 1001;
	}

	required VectorDesc half_extents = 1;
	
}

message SceneSphereDesc {

  extend SceneObjectDesc {
		required SceneSphereDesc scene_object = 1002;
	}

	required float radius = 1;

}

message SceneLightDesc {

  extend SceneObjectDesc {
		required SceneLightDesc scene_object = 2000;
	}

	required VectorDesc ambient  = 1;
	required VectorDesc diffuse  = 2;
	required VectorDesc specular = 3;
}


message ConstraintDesc {

  extensions 1000 to max;

	enum Type {
    HINGE = 0;
    UNIVERSAL = 1;
	}

	required Type type = 1;
  required string name = 2;

	// The bodies for the constraint (one or two)
	repeated string body = 3;

}

message HingeConstraintDesc {

  extend ConstraintDesc {
    required HingeConstraintDesc constraint = 1000;
	}

	required VectorDesc pivot_in_a = 1;
	required VectorDesc pivot_in_b = 2;
	required VectorDesc axis_in_a = 3;
	required VectorDesc axis_in_b = 4;
	required VectorDesc limit = 5;

}


message UniversalConstraintDesc {

  extend ConstraintDesc {
    required UniversalConstraintDesc constraint = 1001;
	}

  required VectorDesc pivot = 1;
  required VectorDesc axis_0 = 2;
  required VectorDesc axis_1 = 3;
  required VectorDesc limit_0 = 4;
	required VectorDesc limit_1 = 5;

}


message SceneCreatureDesc {

  extend SceneObjectDesc {
		required SceneCreatureDesc scene_object = 1100;
	}

  required string name = 1;

  repeated SceneObjectDesc child_scene_object = 2;
  repeated ConstraintDesc constraint = 3;

}




